test/tests/functional/pbs_power_provisioning_cray.py

```diff
@@ -449,15 +500,536 @@ def test_last_used_time(self):
+        t = int(time.time())
+        self.logger.info("Waiting for 12 mins to ramp down all the nodes")
+        time.sleep(720)
+        for n in self.server.status(NODE):
+            if 'resources_available.PBScraynid' in n:
+                self.server.expect(NODE, {'state': 'sleep'}, id=n['id'])
+        self.cleanup_power_ramp_rate()
+
+    @timeout(1000)
+    def test_power_ramp_down_max_concurennt_nodes(self):
+        """
+        Test power hook will ramp down only the number of
+        max_concurrent nodes specified in conf file per hook run.
+        """
+        self.setup_power_ramp_rate()
+        a = {"power_ramp_rate_enable": True, 'node_idle_limit': '10',
+             'max_concurrent_nodes': '2'}
+        self.modify_hook_config(attrs=a, id='PBS_power')
+        a = {'freq': 60}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        a = {'enabled': 'True'}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        t = int(time.time())
+        self.logger.info("Waiting for 90 secs to ramp down 2 nodes")
+        time.sleep(90)
+        a = {'enabled': 'False'}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        i = 0
+        for n in self.server.status(NODE):
+            if 'resources_available.PBScraynid' in n:
+                if n['state'] == 'sleep':
+                    i = i + 1
+        print i
+        if i != 2:
+            self.assertFalse("2 nodes are not in sleep state")
+        self.cleanup_power_ramp_rate()
+
+    @timeout(1500)
+    def test_power_ramp_up_nodes(self):
+        """
+        Test when a job is calandered on a vnode which is in sleep state,
+        the node will be ramped up and job will run.
+        """
+        self.setup_power_ramp_rate()
+        self.scheduler.set_sched_config({'strict_ordering': 'True ALL'})
+        a = {"power_ramp_rate_enable": True, 'node_idle_limit': '30'}
+        self.modify_hook_config(attrs=a, id='PBS_power')
+        a = {'freq': 60}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        a = {'enabled': 'True'}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        t = int(time.time())
+        self.logger.info("Waiting for 12 mins to ramp down all the nodes")
+        time.sleep(720)
+        self.server.expect(NODE, {'state': 'sleep'}, id=self.names[0])
+        a = {"node_idle_limit": "1800", 'min_node_down_delay': '30'}
+        self.modify_hook_config(attrs=a, id='PBS_power')
+        jid = self.submit_job(1000, {'Resource_List.vnode': self.names[0],
+                                     'Resource_List.vntype': 'cray_compute'})
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
```

Comment:
Consider doing a self.scheduler.log_match() for 'Job is a top job and will run at'.  That assures your job is calendared.  (and for other calendared jobs in the test plan)